feat: custom Registry port & fallbacks, spawn var fix

dtbuchholz opened the pull request:
## Summary

Tests will occasionally fail without insight, causing a mocha "exceeded timeout" error. With this, a `fallback` and `registryPort` option can now be set, allowing you to enable fallbacks and/or use a custom Registry port for hardhat. Also, the Registry spawn process wasn't using the passed env variables, which could be causing some errors.

Closes [LOT-6](https://linear.app/tableland/issue/LOT-6/ci-improvement-with-tests).

## Details

Port 8545 being already in use—which is what the Registry / hardhat node bind to—seems causes CI errors since hardhat is trying to connect to a taken port. This adds functionality for retry logic that probes port 8545 before spinning up the Registry.

- Created utils to probe / check port availability.
- Allow a `fallback` option (e.g., `npx local-tableland --fallback`) such that if the default port used is in use, it will attempt backup ports and spin up the Registry from there. If so, it'll log something like `[Notice] Registry default port in use, using fallback port 8546`.
- Allow a `registryPort` option (e.g.,  `npx local-tableland --fallback --registryPort 1234`) to set a custom port for hardhat. It'll log something like `[Notice] Registry is using custom port 1234`.
- Altered `getAccounts()` helper to, optionally, take an instance of LT to grab the port being used, in case a fallback occurs. (Will default to using port 8545 if not LT instance is passed, ensuring backwards compatability).
- Created a `getRegistryPort()` helper to get the Registry port number being used in case of fallbacks, also optionally taking LT as a param.
- Stored Registry spawn process variables to `process.env.HARDHAT_...` vs. defining/passing them directly to fix undefined vars being passed.
- Also added some additional error handling to, hopefully, improve insight into what happens when tests fail unexpectedly in mocha before/after hooks.

When you pass a custom port, or if a fallback occurs, the `validator/config.json` is temporarily overwritten. This allows the validator to use whatever port is set. With shutdown processes, this file gets reset to the original, ensuring port changes that happen during execution always return to the original state (i.e., port 8545) and don't cause side effects.

One callout is that you can't run two LT instances at the same time because they're sharing resources. The second instance will error out with something like `initializing db connection: creating migration: disk I/O error: no such file or directory`. But, this base functionality should enable future features like spinning up multiple hardhat nodes on different ports with a single validator, mimicking a multichain testing setup.

## How it was tested

- Added tests for general start and shutdown processes.
- Added tests to check how LT startup happens in the port in use scenario. E.g., if the port is taken for the _all_ of the retry attempts + fails, or if the port is taken temporarily and successfully triggers retries + actually deploys the Registry.
- Added tests for using a custom `registryPort`.
- Also created a rudimentary way to log the `start()` and `shutdown()` process execution times since they showed a high variance, particularly, for shutdown. Sometimes, its average/median was ~10 ms, but often, it's 10 seconds.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

dtbuchholz commented on src/main.ts:
```diff
@@ -73,6 +74,21 @@ class LocalTableland {
     // TODO: I don't think this is doing anything anymore...
     this.#_cleanup();
 
+    // check if the port is in use and retry 5 times (1 second between each)
```
i figured no need for something like exponential backoff...simple probing every 1 second. can adjust if needed.

dtbuchholz commented on src/util.ts:
```diff
@@ -319,3 +320,22 @@ export const getAccounts = function (): Wallet[] {
   });
 };
+
+export function checkPortInUse(port: number): Promise<boolean> {
```
return true/false if the port is in use via checking for `EADDRINUSE`

dtbuchholz commented on test/cli.test.mjs:
```diff
@@ -1,61 +1,77 @@
 const expect = chai.expect;
 
 describe("Project builder", function () {
+  // Store absolute paths to the current working directory & path to tmp dir
```
these changes aren't required, but when doing local-only GH `test.yml` checks with `act`, i ran into issues with its usage of docker + relative paths. i just make the paths absolute.

dtbuchholz commented on test/e2e.test.mjs:
```diff
@@ -6,10 +6,81 @@ import {
+  let server;
+  const lt = new LocalTableland({ silent: true });
+
+  this.timeout(20000); // Starting up LT takes 3000-7000ms; shutting down takes <10-10000ms
```
i observed that on my machine, the LT startup usually takes 3-7 seconds, and shutdown was <10ms to 10 seconds. i'm not sure why the variance. 

e.g., one test would log:
```
  Validator and Chain startup and shutdown
lt start took 4313 milliseconds.
    ✔ fails to start due to being blocked on port 8545 (11328ms)
stopMockServer took 0 milliseconds.
lt.shutdown took 8 milliseconds.
lt start took 6968 milliseconds.
    ✔ successfully starts using retry logic after initially blocked (6968ms)
stopMockServer took 0 milliseconds.
lt.shutdown took 7 milliseconds.
```
and running it again would log:
```
  Validator and Chain startup and shutdown
lt start took 3371 milliseconds.
    ✔ fails to start due to being blocked on port 8545 (10384ms)
stopMockServer took 0 milliseconds.
lt.shutdown took 9 milliseconds.
lt start took 5846 milliseconds.
    ✔ successfully starts using retry logic after initially blocked (5847ms)
stopMockServer took 0 milliseconds.
lt.shutdown took 10021 milliseconds.
```

dtbuchholz commented on test/e2e.test.mjs:
```diff
@@ -19,12 +90,20 @@ describe("Validator, Chain, and SDK work end to end", function () {
+      await lt.start();
+      await new Promise((resolve) => setTimeout(() => resolve(), 2000));
+    } catch (err) {
+      expect.fail(`failed to start local network: ${err.message}`);
```
i figured adding this additional handling can't hurt

dtbuchholz commented on src/main.ts:
```diff
@@ -73,18 +74,38 @@ class LocalTableland {
+    }
+
+    // You *must* store these in `process.env` to access within the hardhat subprocess
+    process.env.HARDHAT_NETWORK = "hardhat";
```
here, i directly store the variables in process.env. if you don't do this and only pass them in the spawned child process, for some reason, they are never read (i.e., undefined within the hardhat scripts).

dtbuchholz commented on src/main.ts:
```diff
@@ -73,18 +74,38 @@ class LocalTableland {
     this.registry = spawn(
       isWindows() ? "npx.cmd" : "npx",
-      ["hardhat", "node"],
+      ["hardhat", "node", "--port", hardhatPort.toString()], // TODO: use a fallback port on conflicts vs. failing on retries
```
this doesn't change any behavior but at least provides the option to use custom port fallbacks, if desired in the future.

joewagner commented on test/e2e.test.mjs:
```diff
@@ -1,30 +1,317 @@
+
+    // Calculate & log the min, max, median, and average start and shutdown times
+    console.log(`\nExecution metrics`);
+    logMetrics("start()", executionTimes.start);
```
This logging is great!

joewagner commented on src/util.ts:
```diff
@@ -69,28 +92,88 @@ const configDescriptors: ConfigDescriptor[] = [
+   * port over the default port, e.g., connect to `http://127.0.0.1:<fallbackPort>`
+   * instead of port 8545.
+   */
+  fallback?: boolean;
```
Does this work from a practical standpoint?  The chain will start on the fallback port, but how will the Validator know how to connect to the fallback?

For the sake of getting the CI tests to pass it seems easier to just wait until the standard port is available.  If we discover that can't work, we could create a non-standard validator config and retry starting with that and the `registryPort` options set.

joewagner commented on src/util.ts:
```diff
@@ -69,28 +92,88 @@ const configDescriptors: ConfigDescriptor[] = [
+   * Use a custom Registry hardhat port, e.g., `http://127.0.0.1:<registryPort>`.
+   * Must have fallbacks enabled.
+   */
+  registryPort?: number;
```
In general enabling a non-default registry port seems great!  Seems worth documenting that you'll have to use a non-default Validator config with the `validator` option. 
ref: https://github.com/tablelandnetwork/local-tableland/pull/423

dtbuchholz replied:
note: created an issue for myself for adding docs on both the `validator` and `registryPort` options: [here](https://linear.app/tableland/issue/DOC-80/update-docs-with-custom-registry-port-and-validator-config)

joewagner commented on .eslintrc.cjs:
```diff
@@ -14,13 +14,21 @@ module.exports = {
+    {
+      files: ["validator/config.json"],
+      rules: {
+        "eol-last": "off",
```
Why turn this off?

joewagner reviewed (changes requested):
@dtbuchholz Nice work here!  
Lots of good changes, however I have a few concerns.  
I added inline comments, but If it's easier I'm happy to jump in a call. 
TL;DR; is two main things:
- the fallback port idea seems like it won't work in practice
- doesn't seem like using try-catch in tests is best practice.  This is certainly up for debate, but we don't do it anywhere else.  If there's a reason it needs to be added here, let's discuss further.

dtbuchholz commented:
@joewagner i've removed the fallback logic and should have your other recommendations incorporated (e.g., removed try/catch from certain test actions, removed unneeded prettier/linting EOL). lmk if i missed anything!

joewagner commented on src/util.ts:
```diff
@@ -69,28 +89,68 @@ const configDescriptors: ConfigDescriptor[] = [
+   */
   docker?: boolean;
+  /**
+   * Use Docker to run the Validator.
```
Should this comment be above the `docker` prop?

joewagner commented on src/up.ts:
```diff
@@ -68,8 +77,12 @@ const go = async function () {
-  console.error("unrecoverable error");
-  console.error(err);
+  if (err.message.includes(`port already in use`)) {
+    console.error(err.message);
```
Sorry for the nit here, but it would be nice to log the entire error object.  It makes debugging unexpected errors more difficult if you don't have a stack trace.

dtbuchholz replied:
i'm all for the nits! updated them.

joewagner reviewed (approved):
@dtbuchholz This is really awesome.  I tested it out with the JS monorepo and it's gonna let us run all the tests in parallel.  That alone is really nice, but if we can also fix the constant CI issues this is a huge win!
Since it's a bug fix I'd like to publish asap.  Sanity check here, does this change any default functionality? i.e. is this a patch version bump?

dtbuchholz commented on src/validators.ts:
```diff
@@ -76,22 +114,37 @@ class ValidatorPkg {
+
+    // reset the Validator config file in case it was modified with a custom
+    // Registry hardhat port
+    if (ORIGINAL_VALIDATOR_CONFIG) {
```
the original file ends with a new line, but the file gets overwritten with no new line char at the end. thus, running LT will show a file has changed, even though it really hasn't. two things come to mind:
- add a `\n` to the end of the file
- use an eol override for eslint and remove it from prettier (i dont _think_ it has a specific file/eol rule i can use)

dtbuchholz commented:
@joewagner nice! and this adds the ability a new feature where you can pass a `--registryPort` flag, but everything else is the same.

one minor detail i just noticed something upon rewriting the validator config, which i _thought_ was a non-issue. see the comment above re: eslint/prettier and an ending return line.

joewagner commented on src/validators.ts:
```diff
@@ -50,6 +58,25 @@ class ValidatorPkg {
+    const validatorConfig = JSON.parse(configFile.toString());
+
+    // save the validator config state
+    ORIGINAL_VALIDATOR_CONFIG = JSON.stringify(validatorConfig, null, 2);
```
@dtbuchholz This is left over from when this package only worked with docker.  What if we change this line to:
`ORIGINAL_VALIDATOR_CONFIG = configFile.toString();`
Then we don't have to deal with changing the lint and prettier setting for all of our ts packages?

joewagner replied:
If that makes sense, we should probably do the same thing on lone 159 too

dtbuchholz replied:
@joewagner yes, nice solution! should be good to go now.

joewagner reviewed (approved):
Excellent!  really excited to push this out to all the tests 🚀
